UHF-8095 paragraphs language support

dire opened the pull request:
# [UHF-8095](https://helsinkisolutionoffice.atlassian.net/browse/UHF-8095)
<!-- What problem does this solve? -->
This is a collection of language support fixes from the ticket.

## What was done
<!-- Describe what was done -->

* `postcss-preset-env` was locked to `^7.0.0`

## How to install

* Make sure your `etusivu `instance is up and running on latest dev branch.
  * `git pull origin dev`
  * `make fresh`
* Update the HDBT theme
  * `composer require drupal/hdbt:dev-UHF-8095_paragraphs-language-support`
* Run `make drush-cr`
* Run `npm run build` inside `hdbt` theme

## How to test
<!-- Describe steps how to test the features, add as many steps as you want to be tested -->
* [x] The file changes should have been checked and tested in each PRs that are merged to this so only quick check should be needed to make sure nothing strange happened in the process.

## Designers review
<!-- One of the checkboxes below needs to be checked like this: `[x]` (or click when not in edit mode) -->

* [x] This PR does not need designers review
* [ ] This PR has been visually reviewed by a designer (Name of the designer)

## Other PRs
<!-- For example an related PR in another repository -->

* Link to other PR


[UHF-8095]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-8095?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Arkkimaagi reviewed (changes requested):
Here is my initial review, more to come once I get the environment working and can properly test this.

Arkkimaagi commented on src/scss/05_objects/_link.scss:
```diff
@@ -28,14 +28,20 @@ a:hover .link__type {
   --external-icon-move: scaleX(var(--lang-direction, 1)) rotate(15deg);
 }
 
+[dir="rtl"] .link__style--highlight {
```
Please also check the linter, these double quotes should not pass linter check.

xkhaven commented on src/scss/06_components/paragraphs/_liftup-with-image.scss:
```diff
@@ -1,5 +1,13 @@
 $breakpoint-design-image-on-left-right: 830px;
 
+[dir="rtl"] .component--liftup-with-image-img {
+  direction: ltr;
```
Is related to positioning image to left or right .. layout will break on RTL direction. Does not affect to body text of that image component.

dire commented on templates/module/helfi_api_base/helfi-link.html.twig:
```diff
@@ -5,7 +5,9 @@
-    {% set attribute_icon %} <span class="link__type link__type--external" aria-label="({{ 'Link leads to external service'|t({}, {'context': 'Explanation for screen-reader software that the icon visible next to this link means that the link leads to an external service.'}) }})"></span>{%- endset %}
+    {% set external_link_id = 'hdbt-link-' ~ random() %}
+    {% set attribute_icon %} <span class="link__type link__type--external" aria-labelledby="{{ external_link_id }}"></span>
+    <span class="is-hidden" id="{{ external_link_id }}">{{ 'Link leads to external service'|t({}, {'context': 'Explanation for screen-reader software that the icon visible next to this link means that the link leads to an external service.'}) }}</span{%- endset %}
```
Nice catch, added.

dire commented on src/scss/06_components/paragraphs/_phasing.scss:
```diff
@@ -36,15 +37,15 @@ $-circle-size--ol-l: $spacing-triple;
     height: calc(100% + $-item-gap + $spacing-half); // Height of the line should overlap 16px with the one below it.
     top: 0;
-    transform: translateX(-50%);
+    transform: translateX(var(--lang-translate-half-start));
```
Switched to use multiplier.

dire commented on src/scss/06_components/paragraphs/_phasing.scss:
```diff
@@ -67,7 +68,7 @@ $-circle-size--ol-l: $spacing-triple;
     height: var(--circle-size);
     top: calc($-item-padding + var(--item-title-lineheight) * 0.5);
-    transform: translate(-50%, -50%);
+    transform: translate(var(--lang-translate-half-start), -50%);
```
Switched to use multiplier.

Arkkimaagi reviewed (changes requested):
The ones with ✅ are ok, the ones with other symbols not so much. :) 

* ⚠️  Text paragraph
    * Quote component needs work, it has aria-label that should be wrapped and border is on wrong side
    * Table cells align to left on RTL language
* ✅  Photo
* ✅  Banner paragraph
* ✅ Remote video
* (Not yet tested) News listing
* ✅ List of links
* ✅ Popular liftup
* ⚠️  Accordion - closing button of accordion has no context and is randomly translated 
![image](https://user-images.githubusercontent.com/1191667/231430093-1769ffbf-a9e3-45f9-b560-1efd59898373.png)
This should be forced to english instead and wrapped properly.
* ⚠️ Sidebar text
    * Bullet points have wrong margin, should be `inline-start`, not `left`.

Arkkimaagi commented on hdbt.theme:
```diff
@@ -1476,3 +1482,22 @@ function hdbt_preprocess_paragraph__school_search(array &$variables): void {
+ */
+function hdbt_preprocess_eu_cookie_compliance_popup_info(array &$variables) {
+  $primaryLanguages = [
+    'fi',
```
Why is this defined second time in same file?

dire replied:
Probably from doing similar things same time.
I believe the `$variables['alternative_language']` from the `hdbt_preprocess()` should be available here to be used in the if statement.

Arkkimaagi commented on templates/content/social-media-links.html.twig:
```diff
@@ -1,6 +1,9 @@
-  {% for element in elements %}
+  {% for key, element in elements %}
+    {# Only email text is translated #}
+    {% set use_lang_attributes = alternative_language and key != 'email' %}
```
I wonder if we should change the email contexts somehow as we can not be 100% sure to which languages email has been translated of all the world languages. Like for example, is it translate to Sami?

dire replied:
Ah maybe it's the "Send as an email" text. Some languages don't have translation for it and then the "lang" and "dir" attributes would be needed.

dire replied:
This is now fixed. If "Send as an email" text is not translated, a `lang `and `dir` attributes are added.

Arkkimaagi commented on templates/misc/embedded-content-cookie-compliance.twig:
```diff
@@ -1,5 +1,5 @@
 <div class="embedded-content-cookie-compliance media-{{ media_id }}">
-  <div class="message">
+  <div class="message" {{ alternative_language ? create_attribute({'lang': 'en', 'dir': 'ltr'}) }}>
```
Should we maybe move the `en` and `ltr` to a common place, so that it's not defined in million places, just in case someone wants to have another default language?

dire replied:
I actually now switched them to be `$variables['lang_attributes']['fallback_lang/fallback_dir]` because if they are inside `['attributes']` they would be added to everything where attributes are added in the twig files. 😬

Arkkimaagi reviewed (changes requested):
A couple of few new points.

Arkkimaagi commented:
Please also fix conflicts with main branch

Jussiles commented on hdbt.theme:
```diff
@@ -1476,3 +1482,22 @@ function hdbt_preprocess_paragraph__school_search(array &$variables): void {
+    'sv',
+  ];
+
+  $language = Drupal::languageManager()->getCurrentLanguage();
```
I think current language should be defined with parameter these days. It should be either interface lang or content lang depending on context:
`getCurrentLanguage(LanguageInterface::TYPE_INTERFACE)`
or
`getCurrentLanguage(LanguageInterface::TYPE_CONTENT)`

In this case I believe it should be content lang but you know better what's happening in this PR :)

Jussiles replied:
Risto has been updating these lately. Check this PR for reference: https://github.com/City-of-Helsinki/drupal-helfi-platform-config/pull/493/files

dire replied:
Good point! I pushed a change to fix this one in the `hdbt_preprocess()`. The commented part of code is already removed so no need fix in that.
I also believe it should be content lang.

dire commented:
> The ones with ✅ are ok, the ones with other symbols not so much. :)
> * ⚠️  Accordion - closing button of accordion has no context and is randomly translated
>   ![image](https://user-images.githubusercontent.com/1191667/231430093-1769ffbf-a9e3-45f9-b560-1efd59898373.png)


I added context for the accordion close button and translations for it in Swedish and Finnish. Random translations shouldn't happen anymore.

dire commented on src/scss/06_components/paragraphs/_list-of-links.scss:
```diff
@@ -202,6 +205,7 @@ $-item-spacing: $spacing-triple;
     .list-of-links__item__title {
       display: flex;
+      margin-inline: 0;
       margin-left: 0;
```
Removed the margin-left: 0.

dire commented on src/scss/06_components/paragraphs/_list-of-links.scss:
```diff
@@ -150,10 +151,12 @@ $-item-spacing: $spacing-triple;
 
     @include breakpoint($breakpoint-s) {
       margin: 0;
+      margin-inline-start: 0;
```
Tested and looks like it doesn't so the margin-inline-start is needed.
Might be because the `margin-inline-start: (100px + $spacing);` (from earlier lines) seems to be under more specific selector `[dir="rtl"] .list-of-links__item .list-of-links__item__title` and the `margin: 0` goes under `.list-of-links__item .list-of-links__item__title`.

But the `margin: 0` is actually not needed so I removed it.

dire commented on src/scss/02_mixins/_tags_mixins.scss:
```diff
@@ -11,6 +11,7 @@ $-tag-interactive--deletable-padding: 12px 20px 12px;
   margin-bottom: 0;
   margin-top: 0;
+  padding-inline-start: 0;
   padding-left: 0;
```
Removed `padding-left`.

dire commented:
> ⚠️ Sidebar text
> Bullet points have wrong margin, should be inline-start, not left.

Looks like this is already fixed.

dire commented:
> * ⚠️  Text paragraph
>   
>   * Quote component needs work, it has aria-label that should be wrapped and border is on wrong side
>   * Table cells align to left on RTL language

The table cells have been fixed now by using `text-align: start`. Border has also been moved to correct side on RTL.

The aria-label needs to be fixed separately in platform config as the code adding it is there.

dire commented:
> Please also fix conflicts with main branch, I think we could remove the dist from this branch totally as it's causing conflicts unnecessarily.

Fixed. Please do a brief sanity check.
Also same time I reverted the changes to the `dist/style.min.css`.

teroelonen reviewed (approved):
This looks good now 🦖
